x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java

```diff
@@ -35,96 +37,182 @@
         QueryFieldsResolver queryFieldsResolver
     ) {
-        this.relevanceSettingsService = relevanceSettingsService;
-        this.curationsService = curationsService;
-        this.queryFieldsResolver = queryFieldsResolver;
+        this.organicQueryRewriter = new OrganicQueryRewriter(clusterService, relevanceSettingsService, queryFieldsResolver);
+        this.curationQueryRewriter = new CurationQueryRewriter(clusterService, curationsService);
     }
 
-    public Query rewriteQuery(RelevanceMatchQueryBuilder relevanceMatchQueryBuilder, SearchExecutionContext context) throws IOException {
-        Map<String, Float> fieldsAndBoosts = retrieveFieldsAndBoosts(relevanceMatchQueryBuilder.getRelevanceSettingsId(), context);
-        final QueryBuilder combinedFieldsBuilder = new CombinedFieldsQueryBuilder(relevanceMatchQueryBuilder.getQuery(), fieldsAndBoosts);
+    public QueryBuilder rewriteQuery(RelevanceMatchQueryBuilder relevanceMatchQuery, SearchExecutionContext context) throws IOException {
+        final QueryBuilder organicQuery = organicQueryRewriter.rewriteQuery(null, relevanceMatchQuery, context);
+        return curationQueryRewriter.rewriteQuery(organicQuery, relevanceMatchQuery, context);
+    }
 
-        return applyCurations(combinedFieldsBuilder, relevanceMatchQueryBuilder).toQuery(context);
+    private interface QueryRewriter {
+        QueryBuilder rewriteQuery(QueryBuilder baseQuery, RelevanceMatchQueryBuilder relevanceMatchQuery, SearchExecutionContext context);
     }
 
-    private static QueryBuilder applyExcludedDocs(QueryBuilder queryBuilder, CurationSettings curationSettings) {
-        if (curationSettings.excludedDocs().isEmpty() == false) {
-            BoolQueryBuilder booleanQueryBuilder = new BoolQueryBuilder();
-            booleanQueryBuilder.should(queryBuilder);
+    private static abstract class AbstractQueryRewriter<S extends Settings> implements QueryRewriter {
+        private final ClusterService clusterService;
+        private final SettingsService<S> settingsService;
+
+        public AbstractQueryRewriter(ClusterService clusterService, SettingsService<S> settingsService) {
+            this.clusterService = clusterService;
+            this.settingsService = settingsService;
+        }
 
-            BoolQueryBuilder excludedDocsBuilder = new BoolQueryBuilder();
-            for (CurationSettings.DocumentReference excludedDoc : curationSettings.excludedDocs()) {
-                BoolQueryBuilder mustQueryBuilder = new BoolQueryBuilder();
-                mustQueryBuilder.must(new TermsQueryBuilder("_index", excludedDoc.index()));
-                mustQueryBuilder.must(new TermsQueryBuilder("_id", excludedDoc.id()));
+        protected S getSettings(RelevanceMatchQueryBuilder relevanceMatchQuery, SearchExecutionContext context) {
+            String settingsId = getSettingsId(relevanceMatchQuery);
+            SearchEngine searchEngine = getSearchEngine(context);
```

Comment:
Would it make sense to quickly check for the existence of a searchEngine first and return if it exists, _then_ check for settings/get settings? That may make it clearer that the existence of a searchEngine is a prerequisite check.